test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java

```diff
@@ -1026,6 +1052,18 @@ protected Query rewrite(Query query) throws IOException {
         return query;
     }
 
+    protected void checkWarningHeaders(String... messages) {
+        final List<String> warnings = threadContext.getResponseHeaders().get(DeprecationLogger.DEPRECATION_HEADER);
+        assertThat(warnings, hasSize(messages.length));
+        for (String msg : messages) {
+            assertThat(warnings, hasItem(equalTo(msg)));
+        }
+        // "clear" current warning headers by setting a new ThreadContext
+        DeprecationLogger.removeThreadContext(this.threadContext);
```

Comment:
I think that we are leaking a thread local here? We should close the current threadContext before overriding it.